index.js

```diff
@@ -1,23 +1,52 @@
 /**
- * @typedef {import('mdast').Root} Root
- * @typedef {import('retext-smartypants').Options} Options
+ * @typedef {import("mdast").Root} Root
+ * @typedef {import("retext-smartypants").Options} Options
  */
 
 import { retext } from "retext";
 import { visit } from "unist-util-visit";
 import smartypants from "retext-smartypants";
 
 /**
- * remark plugin to implement SmartyPants.
+ * Remark plugin to implement SmartyPants.
  *
- * @type {import('unified').Plugin<[Options?] | void[], Root>}
+ * @type {import("unified").Plugin<[Options?] | void[], Root>}
  */
 export default function remarkSmartypants(options) {
-  const processor = retext().use(smartypants, options);
-  const transformer = (tree) => {
+  const processor = retext().use(smartypants, {
+    ...options,
+    // Do not replace ellipses, dashes, backticks because they change string
+    // length, and we couldn't guarantee right splice of text in second visit of
+    // tree
+    ellipses: false,
+    dashes: false,
+    backticks: false,
+  });
+
+  const processor2 = retext().use(smartypants, {
+    ...options,
+    // Do not replace quotes because they are already replaced in the first
+    // processor
+    quotes: false,
+  });
+
+  return (tree) => {
+    let allText = "";
+    let startIndex = 0;
+
+    visit(tree, "text", (node) => {
+      allText += node.value;
+    });
```

Comment:
"inlineCode" doesn't have its text content as "text" children but just has it on value field

```suggestion
    const checkFn = (node) => node.type === "text" || node.type === "inlineCode";
    visit(tree, checkFn, (node) => {
      allText += node.value;
    });
```
same for the second visit.

Reply:
we can just use `['text', 'inlineCode']`